packages/trackerless-network/src/logic/StreamrNode.ts

```diff
@@ -0,0 +1,406 @@
+        const discoveryResult = await this.streamEntryPointDiscovery!.discoverEntryPointsFromDht(
+            streamPartId,
+            knownEntryPoints.length,
+            forwardingPeer
+        )
+        entryPoints = knownEntryPoints.concat(discoveryResult.discoveredEntryPoints)
+        await Promise.all(sampleSize(entryPoints, 4).map((entryPoint) => layer1.joinDht(entryPoint)))
+        await this.streamEntryPointDiscovery!.storeSelfAsEntryPointIfNecessary(
+            streamPartId,
+            discoveryResult.joiningEmptyStream,
+            discoveryResult.entryPointsFromDht,
+            entryPoints.length
+        )
+    }
+
+    private createStream(streamPartID: string, entryPoints: PeerDescriptor[]): [DhtNode, RandomGraphNode] {
+        const layer1 = this.createLayer1Node(streamPartID, entryPoints)
+        const layer2 = this.createRandomGraphNode(streamPartID, layer1)
+        this.streams.set(streamPartID, {
+            type: StreamNodeType.RANDOM_GRAPH,
+            layer1,
+            layer2
+        })
+        layer2.on('message', (message: StreamMessage) => {
+            this.emit('newMessage', message)
+        })
+        return [layer1, layer2]
+    }
+
+    private createLayer1Node = (streamPartID: string, entryPoints: PeerDescriptor[]) => {
+        return new DhtNode({
+            transportLayer: this.layer0!,
+            serviceId: 'layer1::' + streamPartID,
+            peerDescriptor: this.layer0!.getPeerDescriptor(),
+            routeMessageTimeout: 5000,
+            entryPoints: entryPoints,
+            numberOfNodesPerKBucket: 4,
+            rpcRequestTimeout: 15000,
+            dhtJoinTimeout: 60000,
+            nodeName: this.config.nodeName + ':layer1'
+        })
+    }
+
+    private createRandomGraphNode = (streamPartID: string, layer1: DhtNode) => {
+        return createRandomGraphNode({
+            randomGraphId: streamPartID,
+            P2PTransport: this.P2PTransport!,
+            layer1: layer1,
+            connectionLocker: this.connectionLocker!,
+            ownPeerDescriptor: this.layer0!.getPeerDescriptor(),
+            minPropagationTargets: this.config.streamPartitionMinPropagationTargets,
+            numOfTargetNeighbors: this.config.streamPartitionNumOfNeighbors,
+            nodeName: this.config.nodeName,
+            acceptProxyConnections: this.config.acceptProxyConnections
+        })
+    }
+
+    async waitForJoinAndPublish(
+        streamPartId: string,
+        msg: StreamMessage,
+        timeout?: number
+    ): Promise<number> {
+        if (this.getStream(streamPartId)?.type === StreamNodeType.PROXY) {
+            return 0
+        }
+        await this.joinStream(streamPartId)
+        if (this.getStream(streamPartId)!.layer1!.getBucketSize() > 0) {
+            const neighborCounter = new NeighborCounter(this.getStream(streamPartId)!.layer2 as RandomGraphNode, 1)
+            await neighborCounter.waitForTargetReached(timeout || 5001)
```

Comment:
If timeout is` 0` it shouldn't wait?

```suggestion
            await neighborCounter.waitForTargetReached(timeout ?? 5001)
```

Similar also in `waitForJoinAndSubscribe` and multiple other places where we have a default for a number.